examples/basic_authentication/lib/user_repository.dart

```diff
@@ -0,0 +1,128 @@
+import 'dart:convert';
+
+import 'package:crypto/crypto.dart';
+import 'package:equatable/equatable.dart';
+import 'package:meta/meta.dart';
+
+final List<User> _db = [];
+
+@visibleForTesting
+
+/// Used to mock the database of the repository.
+void mockUserData(List<User> users) {
+  _db
+    ..clear()
+    ..addAll(users);
+}
+
+@visibleForTesting
+
+/// Used in tests to get the data from the memory database.
+List<User> getDb() => _db;
+
+/// {@template user}
+/// A User.
+/// {@endtemplate}
+class User extends Equatable {
+  /// {@macro user}
+  const User({
+    required this.id,
+    required this.name,
+    required this.username,
+    required this.password,
+  });
+
+  /// The user's id.
+  final String id;
+
+  /// The user's name.
+  final String name;
+
+  /// The user's username.
+  final String username;
+
+  /// The user's password in a hashed form.
+  final String password;
+
+  @override
+  List<Object?> get props => [id, name, username, password];
+}
+
+/// {@template user_repository}
+/// Repository which manages users.
+/// {@endtemplate}
+class UserRepository {
+  String _hashValue(String value) {
+    return sha256.convert(utf8.encode(value)).toString();
+  }
+
+  /// Checks in the database for a user with the given [username] and
+  /// [password].
+  ///
+  /// The received password should be in plain text, and will be hashed, so it
+  /// can be compared to the stored password hash.
+  Future<User?> userFromCredentials(String username, String password) async {
+    final hashedPassword = _hashValue(password);
+    final users = _db.where(
+      (user) => user.username == username && user.password == hashedPassword,
+    );
```

Comment:
This statement would become more efficient and readable if we consider `_db` as a Map (or HashMap precisely).

Reply:
Related to https://github.com/VeryGoodOpenSource/dart_frog/pull/715#discussion_r1239558068.

Reply:
Again, this is a fictional database, perfomance is not a concern here.

Reply:
Shifting conversation to https://github.com/VeryGoodOpenSource/dart_frog/pull/715#discussion_r1243556038.